protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,521 @@
+    function notifyDistribution(uint256 amount) external override {
+        if (msg.sender != address(rewardToken) && !isElevatedAccess(msg.sender, msg.sig)) { revert CommonEventsAndErrors.InvalidAccess(); }
+        /// @notice Temple Gold contract mints TGLD amount to contract before calling `notifyDistribution`
+        nextRewardAmount += amount;
+        emit GoldDistributionNotified(amount, block.timestamp);
+    }
+    
+    /**  
+     * @notice Get reward data
+     * @return Reward data
+     */
+    function getRewardData() external override view returns (Reward memory) {
+        return rewardData;
+    }
+
+    function _getReward(address _staker, address _to) internal {
+        uint256 amount = claimableRewards[_staker];
+        if (amount > 0) {
+            claimableRewards[_staker] = 0;
+            rewardToken.safeTransfer(_to, amount);
+            emit RewardPaid(_staker, _to, amount);
+        }
+    }
+
+    function _withdrawFor(
+        address _staker,
+        address _to,
+        uint256 _amount,
+        bool _claimRewards,
+        address _rewardsTo
+    ) private updateReward(_staker) {
+        if (_amount == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        uint256 _balance = _balances[_staker];
+        if (_amount > _balance) 
+            { revert CommonEventsAndErrors.InsufficientBalance(address(stakingToken), _amount, _balance); }
+        unchecked {
+            _balances[_staker] = _balance - _amount;
+        }
+        totalSupply -= _amount;
+        _moveDelegates(delegates[_staker], address(0), _amount);
+
+        stakingToken.safeTransfer(_to, _amount);
+        emit Withdrawn(_staker, _to, _amount);
+
+        // claim reward
+        if (_claimRewards) {
+            // can call internal because user reward already updated
+            _getReward(_staker, _rewardsTo);
+        }
+    }
+
+    function _earned(
+        address _account
+    ) internal view returns (uint256) {
+        return _balances[_account] * (_rewardPerToken() - userRewardPerTokenPaid[_account]) / 1e18 
+            + claimableRewards[_account];
+    }
+
+    function _applyStake(address _for, uint256 _amount) internal updateReward(_for) {
+        totalSupply += _amount;
+        _balances[_for] += _amount;
+        stakeTimes[_for] = block.timestamp;
+        emit Staked(_for, _amount);
+    }
+
+    function _rewardPerToken() internal view returns (uint256) {
+        if (totalSupply == 0) {
+            return rewardData.rewardPerTokenStored;
+        }
+
+        return
+            rewardData.rewardPerTokenStored +
+            (((_lastTimeRewardApplicable(rewardData.periodFinish) -
+                rewardData.lastUpdateTime) *
+                rewardData.rewardRate * 1e18)
+                / totalSupply);
+    }
+
+    function _notifyReward(uint256 amount) private {
+        if (block.timestamp >= rewardData.periodFinish) {
+            rewardData.rewardRate = uint216(amount / rewardDuration);
```

Comment:
gas saving:
storage variable `rewardDuration` is read 5 times from storage in `_notifyReward()`. 
Cache it into memory